[threaded-animations] disable accelerated animations if threaded and non-threaded animations animate the same underlying property

graouts opened the pull request:
#### 28045a5b27fe38ad7b575025db2d810e27bce7c3
<pre>
[threaded-animations] disable accelerated animations if threaded and non-threaded animations animate the same underlying property
<a href="https://bugs.webkit.org/show_bug.cgi?id=302856">https://bugs.webkit.org/show_bug.cgi?id=302856</a>
<a href="https://rdar.apple.com/165120183">rdar://165120183</a>

Reviewed by Simon Fraser.

Until we have monotonic and scroll-driven animations all accelerated in the remote layer tree, we need
to ensure the different accelerated approaches can coexist. This means that in the case where a monotonic
and a scroll-driven animation both target the same underlying property, we need to disable accelerated
animations altogether.

To that end, we extend `KeyframeEffectStack::allowsAcceleration()` to identify such cases and return false.
Then, in `RenderLayerBacking::updateAcceleratedEffectsAndBaseValues()` we simply ignore targets with an
associated effect stack that is marked as such.

To test this is implemented correctly, we add tests that have two animations that can be accelerated specified
on the same element. In one test, these two animations target different properties and thus allow acceleration
with mixed types. In the other test, these two animations target the same underlying property and thus disallow
acceleration of both types.

Tests: webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-different-accelerated-property.html
       webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-dynamic.html
       webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property.html

* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-different-accelerated-property-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-different-accelerated-property.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-dynamic-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-dynamic.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations.html:
* LayoutTests/webanimations/threaded-animations/threaded-animations-utils.js:
* Source/WebCore/animation/AcceleratedEffectStackUpdater.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::effectStackNoLongerAllowsAccelerationDuringAcceleratedActionApplication):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::allowsAcceleration const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):

Canonical link: <a href="https://commits.webkit.org/303444@main">https://commits.webkit.org/303444@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e130245c5c08e2983678512bbe66ef2c3b2123c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132435 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43481 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139950 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84395 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/441f3010-da59-427a-a552-5f6653529fff) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134305 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4689 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101245 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ef2f08a9-0d4e-44f3-b663-f74f765e55a1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135381 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118626 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82038 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e6cb3e48-80d3-4dce-a7a8-0d4ee1f1b456) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83180 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142604 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4600 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37332 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109622 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4682 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3968 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109802 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3484 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114900 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57901 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20571 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4654 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33251 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4488 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68105 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4745 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4611 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -2894,7 +2894,12 @@ void KeyframeEffect::effectStackNoLongerAllowsAccelerationDuringAcceleratedActio
 #if ENABLE(THREADED_ANIMATIONS)
     if (canHaveAcceleratedRepresentation()) {
-        ASSERT_NOT_REACHED();
+        if (RefPtr document = this->document()) {
```
How about putting this code in a lambda that's only called from the assertion, to remove the overhead of getting `document()` and `settings()` in release.

smfr commented on Source/WebCore/animation/KeyframeEffectStack.cpp:
```diff
@@ -209,6 +209,37 @@ void KeyframeEffectStack::effectAbilityToBeAcceleratedDidChange(const KeyframeEf
+// FIXME: We can remove all of this once time-based and progress-based animations are both
+// threaded by default and the associated settings are removed.
+struct AcceleratedPropertyGroup {
+    bool filter { false };
```
Maybe an OptionSet<> instead of 3 bools? Maybe you don't need this struct at all?

graouts committed:
[threaded-animations] disable accelerated animations if threaded and non-threaded animations animate the same underlying property
https://bugs.webkit.org/show_bug.cgi?id=302856
rdar://165120183

Reviewed by Simon Fraser.

Until we have monotonic and scroll-driven animations all accelerated in the remote layer tree, we need
to ensure the different accelerated approaches can coexist. This means that in the case where a monotonic
and a scroll-driven animation both target the same underlying property, we need to disable accelerated
animations altogether.

To that end, we extend `KeyframeEffectStack::allowsAcceleration()` to identify such cases and return false.
Then, in `RenderLayerBacking::updateAcceleratedEffectsAndBaseValues()` we simply ignore targets with an
associated effect stack that is marked as such.

To test this is implemented correctly, we add tests that have two animations that can be accelerated specified
on the same element. In one test, these two animations target different properties and thus allow acceleration
with mixed types. In the other test, these two animations target the same underlying property and thus disallow
acceleration of both types.

Tests: webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-different-accelerated-property.html
       webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-dynamic.html
       webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property.html

* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-different-accelerated-property-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-different-accelerated-property.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-dynamic-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-dynamic.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-on-same-element-for-same-accelerated-property.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations.html:
* LayoutTests/webanimations/threaded-animations/threaded-animations-utils.js:
* Source/WebCore/animation/AcceleratedEffectStackUpdater.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::effectStackNoLongerAllowsAccelerationDuringAcceleratedActionApplication):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::allowsAcceleration const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):

Canonical link: https://commits.webkit.org/303444@main
